File: sdk_src/TCP_Client/MobileIMSDK4a_tcp_Open/app/src/main/java/net/x52im/mobileimsdk/android/ClientCoreSDK.java
/*
 * Copyright (C) 2021  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK_TCP (MobileIMSDK v6.x TCP版) Project. 
 * All rights reserved.
 * 
 * > Github地址：https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址：  http://www.52im.net/forum-89-1.html
 * > 技术社区：  http://www.52im.net/
 * > 技术交流群：215477170 (http://www.52im.net/topic-qqgroup.html)
 * > 作者公众号：“即时通讯技术圈】”，欢迎关注！
 * > 联系作者：  http://www.52im.net/thread-2792-1-1.html
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * ClientCoreSDK.java at 2021-7-1 15:08:17, code by Jack Jiang.
 */
package net.x52im.mobileimsdk.android;

import net.x52im.mobileimsdk.android.core.AutoReLoginDaemon;
import net.x52im.mobileimsdk.android.core.KeepAliveDaemon;
import net.x52im.mobileimsdk.android.core.LocalDataReciever;
import net.x52im.mobileimsdk.android.core.LocalDataSender;
import net.x52im.mobileimsdk.android.core.LocalSocketProvider;
import net.x52im.mobileimsdk.android.core.QoS4ReciveDaemon;
import net.x52im.mobileimsdk.android.core.QoS4SendDaemon;
import net.x52im.mobileimsdk.android.event.ChatBaseEvent;
import net.x52im.mobileimsdk.android.event.ChatMessageEvent;
import net.x52im.mobileimsdk.android.event.MessageQoSEvent;
import net.x52im.mobileimsdk.server.protocal.c.PLoginInfo;

import android.app.Application;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

public class ClientCoreSDK {
    private final static String TAG = ClientCoreSDK.class.getSimpleName();

    public static boolean DEBUG = true;
    public static boolean autoReLogin = true;
    private static ClientCoreSDK instance = null;

    private boolean _init = false;

    //## 20200817备注：localDeviceNetworkOk字段的更新依赖于APP收到的系统的网络事件，而此事件在某些极端情况下
    //##              无法正常收到，进而当网络本身已恢复而此字段并未正确更新的情况下，会导致LocalDataSender中的send
    //##              数据方法在前置检查时，抛出 "ErrorCode.ForC.LOCAL_NETWORK_NOT_WORKING"错误而不会继续数据发送，
    //##              进而无法实现重连指令的发出，从而无法实现重连成功。有鉴于此，停用这个字段是更好的选择！
//	/**
//	 * 网络是否可用, true表示可用，否则表示不可用.
//	 * <p>
//	 * 本字段将在网络事件通知处理中被设置.
//	 * <p>
//	 * 注意：本类中的网络状态变更事件，尤其在网络由断变好之后，受Android系统广播机制的影响事件收到延迟在1~2秒，目
//	 * 前没有找到其它更优的替代方案，但从算法逻辑上讲不影响本核心类库的工作（仅影响核心库算法的构造难易度而已）！
//	 */
//	private boolean localDeviceNetworkOk = true;
    private boolean connectedToServer = true;
    private boolean loginHasInit = false;
    private PLoginInfo currentLoginInfo = null;

    private ChatBaseEvent chatBaseEvent = null;
    private ChatMessageEvent chatMessageEvent = null;
    private MessageQoSEvent messageQoSEvent = null;

    private Context context = null;

    public static ClientCoreSDK getInstance() {
        if (instance == null)
            instance = new ClientCoreSDK();
        return instance;
    }

    private ClientCoreSDK() {
//		init();
    }

    public void init(Context _context) {
        if (!_init) {
            if (_context == null)
                throw new IllegalArgumentException("context can't be null!");

            if (_context instanceof Application)
                this.context = _context;
            else {
                this.context = _context.getApplicationContext();
            }

            // Register for broadcasts when network status changed
            IntentFilter intentFilter = new IntentFilter();
            intentFilter.addAction(ConnectivityManager.CONNECTIVITY_ACTION);
            this.context.registerReceiver(networkConnectionStatusBroadcastReceiver, intentFilter);

            AutoReLoginDaemon.getInstance();
            KeepAliveDaemon.getInstance();
            LocalDataReciever.getInstance();
            QoS4ReciveDaemon.getInstance();
            QoS4SendDaemon.getInstance();

            //
            _init = true;
        }
    }

    public void release() {
        ClientCoreSDK.getInstance().setConnectedToServer(false);// Bug FIX:  2021-02-23 add by Jack Jiang
        LocalSocketProvider.getInstance().closeLocalSocket();
        AutoReLoginDaemon.getInstance().stop(); // 2014-11-08 add by Jack Jiang
        QoS4SendDaemon.getInstance().stop();
        KeepAliveDaemon.getInstance().stop();
//		LocalUDPDataReciever.getInstance().stop();
        QoS4ReciveDaemon.getInstance().stop();

        //## Bug FIX: 20180103 by Jack Jiang START
        QoS4SendDaemon.getInstance().clear();
        QoS4ReciveDaemon.getInstance().clear();
        //## Bug FIX: 20180103 by Jack Jiang END

        try {
            context.unregisterReceiver(networkConnectionStatusBroadcastReceiver);
        } catch (Exception e) {
            Log.i(TAG, "还未注册android网络事件广播的监听器，本次取消注册已被正常忽略哦.");
        }

        _init = false;
        this.setLoginHasInit(false);
        this.setConnectedToServer(false);
    }

    public void setCurrentLoginInfo(PLoginInfo currentLoginInfo) {
		this.currentLoginInfo = currentLoginInfo;
	}

    public PLoginInfo getCurrentLoginInfo() {
		return this.currentLoginInfo;
	}

	public void saveFirstLoginTime(long firstLoginTime) {
		if(this.currentLoginInfo != null)
			this.currentLoginInfo.setFirstLoginTime(firstLoginTime);
	}

    @Deprecated
	public String getCurrentLoginUserId()
	{
		return this.currentLoginInfo.getLoginUserId();
	}

    @Deprecated
	public String getCurrentLoginToken()
	{
		return this.currentLoginInfo.getLoginToken();
	}

    @Deprecated
	public String getCurrentLoginExtra()
	{
		return this.currentLoginInfo.getExtra();
	}

    public boolean isLoginHasInit() {
        return loginHasInit;
    }

    public ClientCoreSDK setLoginHasInit(boolean loginHasInit) {
        this.loginHasInit = loginHasInit;
        return this;
    }

    public boolean isConnectedToServer() {
        return connectedToServer;
    }

    public void setConnectedToServer(boolean connectedToServer) {
        this.connectedToServer = connectedToServer;
    }

    public boolean isInitialed() {
        return this._init;
    }

    public void setChatBaseEvent(ChatBaseEvent chatBaseEvent) {
        this.chatBaseEvent = chatBaseEvent;
    }

    public ChatBaseEvent getChatBaseEvent() {
        return chatBaseEvent;
    }

    public void setChatMessageEvent(ChatMessageEvent chatMessageEvent) {
        this.chatMessageEvent = chatMessageEvent;
    }

    public ChatMessageEvent getChatMessageEvent() {
        return chatMessageEvent;
    }

    public void setMessageQoSEvent(MessageQoSEvent messageQoSEvent) {
        this.messageQoSEvent = messageQoSEvent;
    }

    public MessageQoSEvent getMessageQoSEvent() {
        return messageQoSEvent;
    }

    //--------------------------------------------------------------------------------------- inner class

    private final BroadcastReceiver networkConnectionStatusBroadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            ConnectivityManager connectMgr = (ConnectivityManager) context.getSystemService(context.CONNECTIVITY_SERVICE);
            NetworkInfo mobNetInfo = connectMgr.getNetworkInfo(ConnectivityManager.TYPE_MOBILE);
            NetworkInfo wifiNetInfo = connectMgr.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
            NetworkInfo ethernetInfo = connectMgr.getNetworkInfo(9);
            if (!(mobNetInfo != null && mobNetInfo.isConnected())
                    && !(wifiNetInfo != null && wifiNetInfo.isConnected())
                    // ## Bug FIX 20170228: 解决当Android系统用有线网连接时没有判断此网事件的问题
                    && !(ethernetInfo != null && ethernetInfo.isConnected())
			) {
                Log.w(TAG, "【IMCORE-TCP】【本地网络通知】检测本地网络连接断开了!");
                LocalSocketProvider.getInstance().closeLocalSocket();
            } else {
                if (ClientCoreSDK.DEBUG)
                    Log.i(TAG, "【IMCORE-TCP】【本地网络通知】检测本地网络已连接上了!");
                LocalSocketProvider.getInstance().closeLocalSocket();
            }
        }
    };
}

File: sdk_src/TCP_Client/SwiftMobileIMSDK(ios端swift版,由社区网友开发,仅作学习参考)/SwiftMobileIMSDK/core/ClientCoreSDK.swift
//
//  ClientCoreSDK.swift
//  SwiftMobileIMSDK
//
//  Created by fishbay on 2021/8/1.
//

import Foundation

class ClientCoreSDK: HandyJSON {
    /// debug
    static var enableDebug = false
    /// 是否在登陆成功后掉线时自动重新登陆线程中实质性发起登陆请求
    static var autoReLogin = true
    
    /// 是否已成功连接到服务器（当然，前提是已成功发起过登陆请求后）
    var connectedToServer: Bool = false
    /// 当且仅当用户从登陆界面成功登陆后设置本字段为true，系统退出（登陆）时设置为false
    var loginHasInit: Bool = false
    /// 保存提交到服务端的用户登陆信息，可能是登陆用户名、任意不重复的id等，具体意义由业务层决定
    var currentLoginInfo: LoginInfo?
    /// 是否初始化
    var initial: Bool = false
    /// 网络状态检测类
    var internetReachability: Reachability?
    
    /// 事件回调
    var chatMessageEvent: ChatMessageEvent?
    var chatBaseEvent: ChatBaseEvent?
    var messageQoSEvent: MessageQoSEvent?
    
    /// 单例
    private static let instance: ClientCoreSDK = ClientCoreSDK()
    static func sharedInstance() -> ClientCoreSDK {
        return instance
    }
    required init() {}
    
    /// 设置调试
    /// - Parameter enableDebug: 调试标志
    static func setEnableDebug(enableDebug: Bool) {
        Self.enableDebug = enableDebug
    }
    static func isEnableDebug() -> Bool {
        return enableDebug
    }
    
    /// 设置自动重新登陆
    /// - Parameter autoReLogin: 自动重新登陆
    static func setAutoReLogin(autoReLogin: Bool) {
        Self.autoReLogin = autoReLogin
    }
    static func isAutoReLogin() -> Bool {
        return autoReLogin
    }
    
    /// 初始化核心库
    func initCore() {
        if self.initial {
            // 已初始化，直接返回
            return
        }
        
        self.connectedToServer = false
        self.loginHasInit = false
        
        if self.internetReachability == nil {
            do {
                let reachability = try Reachability()
                NotificationCenter.default.addObserver(self, selector: #selector(reachabilityChanged(notification:)), name: .reachabilityChanged, object: reachability)
                self.internetReachability = reachability
            } catch {
                CAPrint("ClientCoreSDK 创建Reachability对象失败：\(error)")
            }
        }
        
        do {
            try self.internetReachability!.startNotifier()
        } catch {
            CAPrint("ClientCoreSDK本地网络通知启动失败：\(error)")
        }
        
        self.initial = true
        
        CAPrint("ClientCoreSDK已经完成initCore了")
    }
    
    /// 监控网络状态变化
    /// - Parameter notification: 通知对象
    @objc func reachabilityChanged(notification: Notification) {
        
        let reachability = notification.object as! Reachability
        
        switch reachability.connection {
        case .unavailable:
            CAPrint("【IMCORE-TCP】【本地网络通知】检测本地网络连接断开了!")
            LocalSocketProvider.sharedInstance().closeLocalSocket()
            break
        case .wifi:
            CAPrint("【IMCORE-TCP】【本地网络通知】检测本地网络已连接上了! 【wifi】")
            if self.getLoginHasInit() ?? false {
                LocalSocketProvider.sharedInstance().closeLocalSocket()
            }
            
            break
        case .cellular:
            CAPrint("【IMCORE-TCP】【本地网络通知】检测本地网络已连接上了! 【Cellular】")
            if self.getLoginHasInit() ?? false {
                LocalSocketProvider.sharedInstance().closeLocalSocket()
            }
            break
        default:
            break
        }
    }
    
    /// 释放MobileIMSDK框架资源统一方法
    func releaseCore() {
        // 设置是否正常连接（登陆）到服务器的标识
        self.connectedToServer = false
        
        // 尝试停掉掉线重连线程（如果线程正在运行的话）
        AutoReLoginDaemon.sharedInstance().stop()
        
        // 尝试停掉QoS质量保证（发送）心跳线程
        QoS4SendDaemon.sharedInstance().stop()
        
        // 尝试停掉Keep Alive心跳线程
        KeepAliveDaemon.sharedInstance().stop()
        
        // 尝试停掉QoS质量保证（接收防重复机制）心跳线程
        QoS4ReceiveDaemon.sharedInstance().stop()
        
        // 尝试关闭本地Socket
        LocalSocketProvider.sharedInstance().closeLocalSocket()
        
        // 并清除QoS发送队列缓存：防止不退出APP时切换另一账号后qos的缓存队列未清空
        QoS4SendDaemon.sharedInstance().clear()
        
        // 并清除QoS接收队列缓存：防止不退出APP时切换另一账号后qos的缓存队列未清空
        QoS4ReceiveDaemon.sharedInstance().clear()
        
        // 取消注册网络事件的监听
        self.internetReachability?.stopNotifier()
        NotificationCenter.default.removeObserver(self, name: .reachabilityChanged, object: self.internetReachability)
        
        self.initial = false
        self.loginHasInit = false
    }
    
    /// 获取是否初始化状态
    /// - Returns: true - 已经出事后，false - 未初始化
    func isInitialed() -> Bool {
        return self.initial
    }
    
    /// 保存第一次登陆时间
    /// - Parameter firstLoginTime: 第一次登陆时间
    func saveFirstLoginTime(firstLoginTime: Int) {
        self.currentLoginInfo?.firstLoginTime = firstLoginTime
    }
    
    /// 网络是否连接
    /// - Returns: true - 已连接，false - 未连接
    func isInternetReachable() -> Bool {
        let netState = internetReachability?.connection
        return netState == .cellular || netState == .wifi
    }
    
    /// 是否登陆
    /// - Parameter loginHasInit: 是否登陆
    func setLoginHasInit(loginHasInit: Bool?) {
        self.loginHasInit = loginHasInit ?? false
    }
    func getLoginHasInit() -> Bool? {
        return self.loginHasInit
    }
    
    /// 设置当前登陆用户
    /// - Parameter loginInfo: 登陆用户信息
    func setCurrentLoginInfo(loginInfo: LoginInfo) {
        self.currentLoginInfo = loginInfo
    }
    func getCurrentLoginInfo() -> LoginInfo? {
        return self.currentLoginInfo
    }
    
    /// 设置基础事件回调
    /// - Parameter chatBaseEvent: 回调对象
    func setChatBaseEvent(chatBaseEvent: ChatBaseEvent) {
        self.chatBaseEvent = chatBaseEvent
    }
    func getChatBaseEvent() -> ChatBaseEvent? {
        return self.chatBaseEvent
    }
    
    /// 设置数据事件回调
    /// - Parameter chatMessageEvent: 回调对象
    func setChatMessageEvent(chatMessageEvent: ChatMessageEvent) {
        self.chatMessageEvent = chatMessageEvent
    }
    func getChatMessageEvent() -> ChatMessageEvent? {
        return self.chatMessageEvent
    }
    
    /// 设置质量事件回调
    /// - Parameter messageQoSEvent: 回调对象
    func setMessageQoSEvent(messageQoSEvent: MessageQoSEvent) {
        self.messageQoSEvent = messageQoSEvent
    }
    func getMessageQoSEvent() -> MessageQoSEvent? {
        return self.messageQoSEvent
    }
}

